src/nvim/os/env.c

```diff
@@ -346,23 +361,54 @@ void init_homedir(void)
   }
 #endif
 
+  if (var == NULL) {
+    uv_mutex_lock(&mutex);
+    var = os_homedir();
+    uv_mutex_unlock(&mutex);
+  }
+
   if (var != NULL) {
 #ifdef UNIX
     // Change to the directory and get the actual path.  This resolves
     // links.  Don't do it when we can't return.
     if (os_dirname((char_u *)os_buf, MAXPATHL) == OK && os_chdir(os_buf) == 0) {
-      if (!os_chdir(var) && os_dirname(IObuff, IOSIZE) == OK) {
+      if (!os_chdir(var) && os_dirname((char_u *)IObuff, IOSIZE) == OK) {
         var = (char *)IObuff;
       }
       if (os_chdir(os_buf) != 0) {
         EMSG(_(e_prev_dir));
       }
     }
 #endif
-    homedir = xstrdup(var);
   }
+
+  if ((var == NULL || *var == NUL)
+      && os_dirname((char_u *)os_buf, MAXPATHL) == OK) {
+    var = os_buf;
+  }
+
+  homedir = xstrdup(var);
+}
+
+static char homedir_buf[MAXPATHL];
+
+char *os_homedir(void)
+{
+  size_t homedir_size = MAXPATHL;
+  if (uv_os_homedir(os_buf, &homedir_size) == 0) {
```

Comment:
Why not use `homedir_buf` from the beginning?

Reply:
Done